sklearn/utils/estimator_checks.py

```diff
@@ -3297,12 +3296,22 @@ def param_filter(p):
                 tuple,
                 type(None),
                 type,
-                types.FunctionType,
-                joblib.Memory,
             }
             # Any numpy numeric such as np.int32.
             allowed_types.update(np.core.numerictypes.allTypes.values())
-            assert type(init_param.default) in allowed_types, (
+
+            allowed_value = (
+                type(init_param.default) in allowed_types
+                or
+                # Although callables are mutable, we accept them as argument
+                # default value and trust that neither the implementation of
+                # the callable nor of the estimator changes the state of the
+                # callable, which is needed for consistency regarding
+                # pickling and unpickling objects.
+                callable(init_param.default)
+            )
+
+            assert allowed_value, (
```

Comment:
It would be great to change the error message to add that it can be a callable.

Right now the error message only mentions `allowed_types`.

Reply:
I would also maybe remove "All init parameters have to be immutable to make cloning possible"